### PR TITLE
Implementation of functions starting based on loading status & fix problems caused by navigation

### PR DESCRIPTION
**Pull request by samolukadjo**

### CODE REVIEW

**Comment by tusx** [comment]
i will probably get this merged by tomorrow.
🫡

**Comment by samolukadjo** [comment]
I've done some additional work, problems caused by the fact that YouTube doesn't refresh the page when you navigate to different one should be fixed.

If anything else in the future needs to be executed again because of the webpage changing, it can be added to the event listener function.

File: app.js [hunk 1]
```diff
@@ -2,7 +2,7 @@
 // show extension is loaded
 console.log("[Better Mobile Youtube] Better Mobile Youtube has been Loaded!");
 
-
+let currentURL = window.location.href.replace(/\#.*/, '');
```
Comment (tusx) [inline app.js:5]: can you please explain why did you add this, thanks.

File: app.js [hunk 2]
```diff
@@ -140,7 +156,18 @@ function on_video_played() {
...
+    // Check whether the settings button exists since firefox prevents it from loading due to autoplay with sound prevention
+    function loadCheckSettings() {
+        if (document.querySelector('.player-settings-icon')) {
+            on_player_settings_clicked()
+        } else {
+            setTimeout(loadCheckSettings, 25);
+        }
+    }
+    
+    loadCheckSettings()
```
Comment (tusx) [inline app.js]: can this be moved into the run_extension_functions() 
just so that it be easy to know which functions run on when extention starts

File: app.js [hunk 3]
```diff
@@ -153,12 +180,18 @@ function on_video_played() {
...
+        if (videoElement.checkVisibility()) {
+            console.log("[Better Mobile Youtube] Player is visible");
+            on_video_played(videoElement);           
+        } else {
+            setTimeout(loop_to_check_player_visibility, 150);
+            console.log("[Better Mobile Youtube] Player is not visible");
+        }
     } else {
-        setTimeout(loop_to_check_player_visibility, 1000);
+        setTimeout(loop_to_check_player_visibility, 150);
```
Comment (tusx) [inline app.js:191]: just a concern, looping this fast will it cause any issues on low end devices.
the reason i looped ever sec was so under the impression that looping fast may cause high cpu usage on low end devices.

File: app.js [hunk 4]
```diff
@@ -265,27 +298,38 @@ function on_yt_logo_clicked() {
...
 
     // setTimeout(on_video_played, 1000);
 
     // setTimeout(open_link_in_new, 1500);
 
-    setTimeout(loop_to_check_player_visibility, 1000);
+    setTimeout(loop_to_check_player_visibility, 200);
 
 
+    // Check when the url changes and insure that the actual webpage has changed, if yes run loop_to_check_player_visibility again
```
Comment (tusx) [inline app.js:323]: could this be made into a function and then, the function be called inside run_extension_functions()

File: app.js [hunk 5]
```diff
@@ -298,5 +342,16 @@ function run_extension_functions() {
...
-setTimeout(run_extension_functions, 1000); // Delay of 1000 = 1 seconds
+function loadCheck() {
+    if (document.getElementsByClassName('mobile-topbar-header-endpoint')) {
+        run_extension_functions()
+    } else {
+        setTimeout(loadCheck, 15);
+    }
+}
 
+loadCheck()
```
Comment (tusx) [inline app.js]: just to be better readable maybe change the function name to check_page_elements_loaded() or something in those lines so if someone is reading the code could easily understand what this function is doing

**Review by tusx — COMMENTED** [review]
i just added some comments if you could please make the necessary changes so i could merge it.
I was busy irl so was not able to look at the MR.

thanks for all your work.

File: app.js [hunk 6]
```diff
@@ -265,27 +298,38 @@ function on_yt_logo_clicked() {
...
     // setTimeout(on_video_played, 1000);
 
     // setTimeout(open_link_in_new, 1500);
 
-    setTimeout(loop_to_check_player_visibility, 1000);
+    setTimeout(loop_to_check_player_visibility, 200);
 
 
+    // Check when the url changes and insure that the actual webpage has changed, if yes run loop_to_check_player_visibility again
+    window.addEventListener('popstate', function (event) {
```
Comment (tusx) [inline app.js]: could this be made into a function and then, the function be called inside run_extension_functions()

**Comment by tusx** [comment]
and to make the future development better, i have created a dev branch to can keep all the changes there and when a version push will suffice it will be merged into main. 

any thoughts on that matter is welcomed.

File: app.js (hunk 1 above)
Comment (samolukadjo) [reply app.js:5]: This regex gets rid of any # and everything after it, this means that URL changes are only detected when the actual webpage changes, not when a menu or popup is opened.

File: app.js (hunk 2 above)
Comment (samolukadjo) [reply app.js]: I've looked, it seems that moving the function call shouldn't cause problems, and the function itself probably shouldn't have been defined inside another function anyway.

I moved it.

File: app.js (hunk 5 above)
Comment (samolukadjo) [reply app.js]: Thanks

File: app.js (hunk 6 above)
Comment (samolukadjo) [reply app.js]: 👍

File: app.js (hunk 3 above)
Comment (samolukadjo) [reply app.js:191]: I will see about this.

Please wait

Comment (samolukadjo) [reply app.js:191]: After some testing with throttling, I've determined that, at least in this case and those like it, the performance hit is negligible.

Screenshot from 2024-06-29 19-00-39

This is with my laptops CPU throttled to 20 times slower than usual. This function only takes up 0.3-0.5% of the processing time.

While having 50 functions like this one will create a hit that might be noticeable, as long as we keep things reasonable, it should be fine.

**Comment by samolukadjo** [comment]
> and to make the future development better, i have created a dev branch to can keep all the changes there and when a version push will suffice it will be merged into main.
> 
> any thoughts on that matter is welcomed.

This is a good idea, the main branch _is_ supposed to be the stable one that's ready for production or whatnot.

**Comment by samolukadjo** [comment]
I believe that I've addressed your comments, if you have any further ones, please do make them.

**Review by tusx — APPROVED** [review]

**Comment by tusx** [comment]
EVERYTHING LOOKS GOOD I WILL BE MERGING THIS TO MAIN AND THEN to the dev branch
